File: solutions/bit-manipulation/137-single.py
I'll help you implement a solution for the Single Number II problem. This solution will find the number that appears only once in an array where all other numbers appear exactly three times.

```python
#!/usr/bin/env python3
"""
Single Number II - Python Implementation

Problem: Given an array where every element appears three times except for one element
which appears only once, find that single element.

Example:
Input: [2,2,3,2]
Output: 3

Input: [0,1,0,1,0,1,99]
Output: 99
"""

from typing import List


class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        """
        Find the single number that appears only once while others appear three times.
        
        Args:
            nums: List of integers where all numbers except one appear three times
            
        Returns:
            The number that appears only once
            
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        # Initialize ones and twos for bit manipulation
        ones = 0
        twos = 0
        
        for num in nums:
            # First appearance: add to ones
            # Second appearance: remove from ones, add to twos
            # Third appearance: remove from twos
            ones = (ones ^ num) & ~twos
            twos = (twos ^ num) & ~ones
            
        return ones


def test_single_number():
    """Test cases for the singleNumber function."""
    solution = Solution()
    
    # Test case 1
    assert solution.singleNumber([2, 2, 3, 2]) == 3
    
    # Test case 2
    assert solution.singleNumber([0, 1, 0, 1, 0, 1, 99]) == 99
    
    # Test case 3
    assert solution.singleNumber([1]) == 1
    
    # Test case 4
    assert solution.singleNumber([1, 1, 1, 4]) == 4
    
    print("All test cases passed!")


if __name__ == "__main__":
    # Run test cases
    test_single_number()
    
    # Example usage
    solution = Solution()
    example_input = [2, 2, 3, 2]
    result = solution.singleNumber(example_input)
    print(f"Input: {example_input}")
    print(f"Single number: {result}")
```

This implementation includes:

1. A clean and efficient solution using bit manipulation
2. Proper type hints and documentation
3. Clear comments explaining the logic
4. Test cases to verify the implementation
5. Example usage in the main block
6. Proper error handling through the algorithm design

The solution uses a bit manipulation approach that:
- Keeps track of bits that have appeared once (ones) and twice (twos)
- Uses XOR and AND operations to maintain the count of appearances
- Returns the number that appears only once

The time complexity is O(n) where n is the length of the input array, and the space complexity is O(1) as we only use two variables regardless of input size.

The code includes test cases to verify the implementation works correctly for various scenarios and follows Python best practices and conventions.

File: solutions/bit-manipulation/alternatives/137-single.js
/**
 * # Difficulty: Medium
 *
 * # 137. Single Number II
 *
 * Given an integer array nums where every element appears three times except for one, which appears exactly once. Find the single element and return it.
 *
 * You must implement a solution with a linear runtime complexity and use only constant extra space.
 *
 * **Example:**
 *
 * <dl class="example-details">
 * <dt>Input:</dt>
 * <dd>[2,2,3,2]</dd>
 * <dt>Output:</dt>
 * <dd>3 (single number)</dd>
 * <dt>Explanation:</dt>
 * <dd>The single number 3 appears once in [2,2,3,2] (all others appear thrice)</dd>
 * </dl>
 *
 * <details>
 * <summary><b>🔍 SOLUTION EXPLANATION</b></summary>
 *
 * ### INTUITION:
 * [This problem requires understanding of bit manipulation concepts. The key insight is to identify the optimal approach for this specific scenario.]
 *
 * ### APPROACH:
 * 1. **Analyze the problem**: Understand the input constraints and expected output
 * 2. **Choose the right technique**: Apply bit manipulation methodology
 * 3. **Implement efficiently**: Focus on optimal time and space complexity
 * 4. **Handle edge cases**: Consider boundary conditions and special cases
 *
 * ### WHY THIS WORKS:
 * - The solution leverages bit manipulation principles
 * - Time complexity is optimized for the given constraints
 * - Space complexity is minimized where possible
 *
 * ### EXAMPLE WALKTHROUGH:
 * ```
 * Input: nums = [2,2,3,2]
 * Step 1: Count bits at each position
 *   Binary representations:
 *   2 = 010
 *   2 = 010
 *   3 = 011
 *   2 = 010
 *
 *   Bit 0: appears 1 time → 1 % 3 = 1
 *   Bit 1: appears 4 times → 4 % 3 = 1
 *   Bit 2: appears 0 times → 0 % 3 = 0
 *
 * Step 2: Build result from remaining bits
 *   result = 011 (binary) = 3 (decimal)
 *
 * Output: 3 (single number)
 * ```
 *
 * ### TIME COMPLEXITY:
 * O(n)
 *
 * ### SPACE COMPLEXITY:
 * O(1)
 *
 * ### EDGE CASES:
 * - Empty input handling
 * - Single element cases
 * - Large input considerations
 *
 * </details>
 */

/**
 * Main solution for Problem 137: Single Number II
 *
 * @param {number[]} nums - Array where every element appears 3 times except one
 * @return {number} - The single number that appears only once
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */
function solve(nums) {
    // Count bits at each position
    // For numbers appearing 3 times, bit counts are divisible by 3
    // The remainder gives us the single number's bit
    let ones = 0;  // Bits appearing 1 time (mod 3)
    let twos = 0;  // Bits appearing 2 times (mod 3)

    for (const num of nums) {
        // Update twos: add bits that were in ones and are now appearing again
        twos |= ones & num;

        // Update ones: XOR with current number
        ones ^= num;

        // Remove bits that appear 3 times from both ones and twos
        const threes = ones & twos;
        ones &= ~threes;
        twos &= ~threes;
    }

    return ones;
}

/**
 * Test cases for Problem 137: Single
 */
function testSolution() {
    console.log('Testing 137. Single');

    // Test case 1: Basic case
    const result1 = solve([2, 2, 3, 2]);
    const expected1 = 3;
    console.assert(result1 === expected1, `Test 1 failed: expected ${expected1}, got ${result1}`);

    // Test case 2: Single at beginning
    const result2 = solve([0, 1, 0, 1, 0, 1, 99]);
    const expected2 = 99;
    console.assert(result2 === expected2, `Test 2 failed: expected ${expected2}, got ${result2}`);

    // Test case 3: Single element
    const result3 = solve([1]);
    const expected3 = 1;
    console.assert(result3 === expected3, `Test 3 failed: expected ${expected3}, got ${result3}`);

    // Test case 4: Negative numbers
    const result4 = solve([-2, -2, 1, -2]);
    const expected4 = 1;
    console.assert(result4 === expected4, `Test 4 failed: expected ${expected4}, got ${result4}`);

    // Test case 5: Larger array
    const result5 = solve([1, 1, 1, 2, 2, 2, 3]);
    const expected5 = 3;
    console.assert(result5 === expected5, `Test 5 failed: expected ${expected5}, got ${result5}`);

    console.log('All test cases passed for 137. Single!');
}

/**
 * Example usage and demonstration
 */
function demonstrateSolution() {
    console.log('\n=== Problem 137. Single ===');
    console.log('Category: Bit Manipulation');
    console.log('Difficulty: Medium');
    console.log('');

    // Example demonstration would go here
    testSolution();
}

// Run tests if this file is executed directly
if (require.main === module) {
    demonstrateSolution();
}

// Export for use in other modules
module.exports = {
    solve,
    testSolution,
    demonstrateSolution
};

/**
 * Additional Notes:
 * - This solution focuses on bit manipulation concepts
 * - Consider the trade-offs between time and space complexity
 * - Edge cases are crucial for robust solutions
 * - The approach can be adapted for similar problems in this category
 */
